### PR TITLE
Smoldot browser demo deployment

### PR DESCRIPTION
**Pull request by raoulmillais**
## Deploy to GH Pages

* [x] Script build / commit / push to gh-pages branch
* [x] Test deploy to gh-pages - I've deployed it to my fork for testing: https://raoulmillais.github.io/substrate-connect/
* [x] Add documentation to README

## Deploy to IPFS

* [x] Pin dist folder using pianata SDK
    * https://github.com/PinataCloud/Pinata-SDK#pinfromfs
* [x] Unpin old releases
    * https://github.com/PinataCloud/Pinata-SDK#pinList-anchor
    * https://github.com/PinataCloud/Pinata-SDK#unpin-anchor
* [x] Log formatted reponse from pinata
* [x] Log URL constructed from IPFS public gateway and returned content hash
* [x] Add documentation to README

For some reason when served from the IPFS gateway the emojis are garbled.  The HTML content-type is correct though - needs more investigation.

### CODE REVIEW

**Comment by tomaka** [comment]
Here's the quick GitHub Action I put in my repo to deploy github pages: https://github.com/tomaka/test-browser-node/blob/master/.github/workflows/deploy.yml
The IPFS deployment however doesn't work, most likely because the `ipfs-pinata-deploy-action` action is broken.

**Comment by raoulmillais** [comment]
> If that's helpful, here's the quick GitHub Action I put in my repo to deploy github pages: https://github.com/tomaka/test-browser-node/blob/master/.github/workflows/deploy.yml
> The IPFS deployment however doesn't work, most likely because the `ipfs-pinata-deploy-action` action is broken.

Cool thanks... always good to see how others have approached this.  When it comes to deploy pipelines, I prefer to own and maintain my own scripts (which I can run in a simple shell / node runner in an action) myself because:
1. They tend to be reasonably simple
2. They tend to be fairly uniwue with slight differences to every one else's requirements
3. You rarely have to make major changes to them

This leads to  3rd party tools / actions being kitchen sinks trying to cover everyone's requirements so there's more to go wrong and more reason for them to change frequently. I *really never* want my deploy to break because of some upstream dependency change (like the ipfs-deploy-action is broken for you)

**Comment by raoulmillais** [comment]
The PolkadotJS website is deployed to IPFS.  Here's his custom script to pin to Pinata: https://github.com/polkadot-js/apps/blob/master/scripts/ipfsUpload.js  I like this approach (although we won't use cloudflare and a custom domain yet).

**Comment by tomaka** [comment]
> I really never want my deploy to break because of some upstream dependency change 

I don't really see how a custom script fixes that. If the API of the Pinata library changes, it will break both any existing GitHub Action but also any custom script. With a GitHub Action, someone might fix it for you.

**Comment by tomaka** [comment]
For what it's worth, an approach that I like is that it's possible to have GitHub Actions in the repo itself. I don't have an example of someone doing that in the JavaScript world, but [wasmtime does this](https://github.com/bytecodealliance/wasmtime/tree/main/.github/actions). Instead of putting some `.js` file in the repo that someone has to remember to use, they put a `.js` file in some more structure manner.

**Comment by raoulmillais** [comment]
> For what it's worth, an approach that I like is that it's possible to have GitHub Actions in the repo itself. I don't have an example of someone doing that in the JavaScript world, but [wasmtime does this](https://github.com/bytecodealliance/wasmtime/tree/main/.github/actions). Instead of putting some `.js` file in the repo that someone has to remember to use, they put a `.js` file in some more structure manner.



> > I really never want my deploy to break because of some upstream dependency change
> 
> I don't really see how a custom script fixes that. If the API of the Pinata library changes, it will break both any existing GitHub Action but also any custom script. With a GitHub Action, someone might fix it for you.

I guess I'm tainted by past bad experiences with brittle/poorly maintained github actions and tools.  Anyway the crux of my argument is that I dont think this is complicated enough to pull in someone else's scripts

**Comment by raoulmillais** [comment]
> For what it's worth, an approach that I like is that it's possible to have GitHub Actions in the repo itself. I don't have an example of someone doing that in the JavaScript world, but [wasmtime does this](https://github.com/bytecodealliance/wasmtime/tree/main/.github/actions). Instead of putting some `.js` file in the repo that someone has to remember to use, they put a `.js` file in some more structure manner.

I like this a lot :) this is what I hope to do eventually.  Github actions is on my list and I will move my deploy scripts into a similar sort of structure when i get to that (I'll probaly tackle CI and CD together using GH actions)

**Review by Stefie — COMMENTED** [review]
Did you already implement a way to deploy multiple uApps to Guthub pages?
Burnr should theoretically work already

**Comment by raoulmillais** [comment]
> Did you already implement a way to deploy multiple uApps to Guthub pages?
> Burnr should theoretically work already

Not yet.  I don't think it was ready when I opened / started working on this PR.  I just created an issue for that: https://github.com/paritytech/substrate-connect/issues/57

File: package.json [hunk 1]
```diff
@@ -19,6 +19,8 @@
     "build": "yarn workspaces run build",
     "dev:burnr": "yarn workspace @substrate/connect run build && yarn workspace @substrate/burnr run build && yarn workspace @substrate/burnr run start",
     "dev:smoldot-browser-demo": "yarn workspace @substrate/smoldot-browser-demo run dev",
+    "deploy:gh-pages:smoldot-browser-demo": "cd projects/smoldot-browser-demo && yarn deploy:gh-pages",
```
[the remark this answers is not in the document]
Comment (raoulmillais) [reply package.json:22]: I don't think its a good idea to have a "deploy everything everywhere" script in the root of the project.  I'll jot some thoughts down in the issue I created for deploying multiple apps #57

[the remark this answers is not in the document]
Comment (Stefie) [reply package.json:22]: I would just assume that we need to update all apps when we're introducing any significant changes to the `@substrate` packages, there are changes to the `@polkadot` packages, Smoldot or even Substrate, so instead of triggering the updates all manually , there would be one command to deploy all.

[the remark this answers is not in the document]
Comment (raoulmillais) [reply package.json:22]: Let's discuss in #57.

**Comment by raoulmillais** [comment]
TLDR; I'm going to merge this but heads up that when we next update `yarn.lock` there will be some breaking changes coming from `@polkadot/util` which will not result in a compile failure of smoldot-provider but will cause bundling to break. (Time to get on top of that continuous integration ticket to make sure we run `yarn build` on all the projects before merging!)

I merged master into this PR but there was a conflict on `yarn.lock` (as well as an unrelated conflict that I knew how to deal with).  My first attempt to resolve it was delete `yarn.lock` reinstall with `yarn install` and commit the resolved merge. Unfortunately that caused `parcel build` to fail in smoldot-browser-demo  because of changes somewhere in `@polkadot/util`.   This would have probably broken burnr too but I didn't try it.

So to actually merge successfuly I reset to before the merge and resolved it by doing `git checkout --ours` on the `yarn.lock`. Both smoldot-browser-demo and burnr build and dev work fine.
